src/main/scala/edu/berkeley/ce/rockslicing/RockSlicer.scala

```diff
@@ -26,36 +29,49 @@ object RockSlicer {
       System.exit(-1)
     }
     val (globalOrigin, rockVolume, joints) = inputOpt.get
-    var blocks = Seq(Block(globalOrigin, rockVolume))
-    val processorJoints = if (arguments.numProcessors > 1) {
-      LoadBalancer.generateProcessorJoints(blocks.head, arguments.numProcessors)
-    } else {
-      Seq.empty[Joint]
-    }
+    val starterBlocks = Seq(Block(globalOrigin, rockVolume))
 
     // Generate a list of initial blocks before RDD-ifying it
-    if (arguments.numProcessors > 1) {
-      processorJoints foreach { joint =>
-        blocks = blocks.flatMap(_.cut(joint))
+    val seedBlocks = if (arguments.numProcessors > 1) {
+      val processorJoints = LoadBalancer.generateProcessorJoints(starterBlocks.head, arguments.numProcessors)
+      processorJoints.foldLeft(starterBlocks) { (currentBlocks, joint) =>
+        currentBlocks.flatMap(_.cut(joint))
       }
+    } else {
+      starterBlocks
     }
 
-    val blockRdd = sc.parallelize(blocks)
+    val seedBlockRdd = sc.parallelize(seedBlocks)
     val broadcastJoints = sc.broadcast(joints)
 
-    // Iterate through the discontinuities, cutting blocks where appropriate
-    var cutBlocks = blockRdd
-    for (joint <- broadcastJoints.value) {
-      cutBlocks = cutBlocks.flatMap(_.cut(joint, arguments.minRadius, arguments.maxAspectRatio))
+    // Iterate through the discontinuities for each seed block, cutting where appropriate
+    val cutBlocks = seedBlockRdd flatMap { seedBlock =>
+      broadcastJoints.value.zipWithIndex.foldLeft(Seq(seedBlock)) { case (currentBlocks, (joint, idx)) =>
+        if (idx % REDUNDANT_ELIM_FREQ == 0) {
```

Comment:
This is very clever! Would you mind adding some comments to describe what you're doing and why so it'll make sense when we come back to to it later?